metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.kafka.raft.RaftClient;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+import org.apache.kafka.snapshot.SnapshotWriter;
+import org.slf4j.Logger;
+
+import java.util.Optional;
+
+public class SnapshotEmitter implements SnapshotGenerator.Emitter {
+    private final static int DEFAULT_BATCH_SIZE = 100;
+
+    public static class Builder {
+        private int nodeId = 0;
+        private RaftClient<ApiMessageAndVersion> raftClient = null;
+        private int batchSize = DEFAULT_BATCH_SIZE;
+
+        public Builder setNodeId(int nodeId) {
+            this.nodeId = nodeId;
+            return this;
+        }
+
+        public Builder setRaftClient(RaftClient<ApiMessageAndVersion> raftClient) {
+            this.raftClient = raftClient;
+            return this;
+        }
+
+        public Builder setBatchSize(int batchSize) {
+            this.batchSize = batchSize;
+            return this;
+        }
+
+        public SnapshotEmitter build() {
+            if (raftClient == null) throw new RuntimeException("You must set the raftClient.");
+            return new SnapshotEmitter(nodeId,
+                    raftClient,
+                    batchSize);
+        }
+    }
+
+    /**
+     * The slf4j logger to use.
+     */
+    private final Logger log;
+
+    /**
+     * The RaftClient to use.
+     */
+    private final RaftClient<ApiMessageAndVersion> raftClient;
+
+    /**
+     * The maximum number of records to put in each batch.
+     */
+    private final int batchSize;
+
+    private SnapshotEmitter(
+            int nodeId,
+            RaftClient<ApiMessageAndVersion> raftClient,
+            int batchSize
+    ) {
+        this.log = new LogContext("[SnapshotEmitter id=" + nodeId + "] ").logger(SnapshotEmitter.class);
+        this.raftClient = raftClient;
+        this.batchSize = batchSize;
+    }
+
+    @Override
+    public void emit(MetadataImage image) {
+        MetadataProvenance provenance = image.provenance();
+        Optional<SnapshotWriter<ApiMessageAndVersion>> snapshotWriter =
+            raftClient.createSnapshot(provenance.offset(),
+                    provenance.epoch(),
+                    provenance.lastContainedLogTimeMs());
+        if (!snapshotWriter.isPresent()) {
+            log.error("Not generating {} because it already exists.", provenance.snapshotName());
+            return;
+        }
+        try (RaftSnapshotWriter writer = new RaftSnapshotWriter(snapshotWriter.get(), batchSize)) {
+            image.write(writer, new ImageWriterOptions.Builder().
+                    setMetadataVersion(image.features().metadataVersion()).
+                    build());
+            writer.close(true);
+        } catch (Throwable e) {
+            log.error("Encountered error while writing {}", provenance.snapshotName(), e);
+            throw e;
+        } finally {
+            try {
+                snapshotWriter.get().close();
```

Comment:
This would be clearer without using the try-with-resources, which would also end up closing the writer. I do not know which logic will execute first. Perhaps it's best not to depend on the order.

Reply:
Yeah, that's a fair point. I'm sure that this ordering is defined by Java, but it's a bit esoteric.

I was assuming that the close order didn't matter, but perhaps it does (or will eventually?) Best just to be straightforward about it. Since they're both AutoCloseable, I will also use `Utils.closeQuietly` as well.